map<T, double> isn't handled properly

```
error[E0277]: the trait bound `f64: Hash` is not satisfied
  |
5 | #[derive(Debug, Clone, conjure_object::private::Educe)]
  |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Hash` is not implemented for `f64`
```

We'll need some custom logic to handle this case as well.

--- a/changelog/@unreleased/pr-302.v2.yml
+++ b/changelog/@unreleased/pr-302.v2.yml
@@ -1,0 +1,5 @@
+type: fix
+fix:
+  description: Fix codegen for map<T, double>
+  links:
+  - https://github.com/palantir/conjure-rust/pull/302

--- a/conjure-codegen/src/context.rs
+++ b/conjure-codegen/src/context.rs
@@ -856,7 +856,8 @@ impl Context {
             Type::Primitive(PrimitiveType::Double) => true,
             Type::Optional(def) => self.is_double(def.item_type()),
             Type::List(def) => self.is_double(def.item_type()),
-            Type::Primitive(_) | Type::Set(_) | Type::Map(_) | Type::Reference(_) => false,
+            Type::Map(def) => self.is_double(def.value_type()),
+            Type::Primitive(_) | Type::Set(_) | Type::Reference(_) => false,
             Type::External(def) => self.is_double(def.fallback()),
         }
     }

--- a/conjure-object/src/private.rs
+++ b/conjure-object/src/private.rs
@@ -15,6 +15,7 @@ pub use educe::Educe;
 use ordered_float::OrderedFloat;
 use serde::de::{self, IntoDeserializer};
 use std::cmp::Ordering;
+use std::collections::BTreeMap;
 use std::hash::{Hash, Hasher};
 use std::marker::PhantomData;
 use std::{fmt, mem};
@@ -129,6 +130,81 @@ where
         for v in self {
             v.hash(hasher);
         }
+    }
+}
+
+impl<K, V> DoubleOps for BTreeMap<K, V>
+where
+    K: Eq + Ord + Hash,
+    V: DoubleOps,
+{
+    #[inline]
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.iter()
+            .map(|(k, v)| (k, DoubleOpsWrapper(v)))
+            .cmp(other.iter().map(|(k, v)| (k, DoubleOpsWrapper(v))))
+    }
+
+    #[inline]
+    fn eq(&self, other: &Self) -> bool {
+        self.iter()
+            .map(|(k, v)| (k, DoubleOpsWrapper(v)))
+            .eq(other.iter().map(|(k, v)| (k, DoubleOpsWrapper(v))))
+    }
+
+    #[inline]
+    fn hash<H>(&self, hasher: &mut H)
+    where
+        H: Hasher,
+    {
+        self.len().hash(hasher);
+        for (k, v) in self {
+            (k, DoubleOpsWrapper(v)).hash(hasher);
+        }
+    }
+}
+
+struct DoubleOpsWrapper<'a, T>(&'a T);
+
+impl<T> PartialEq for DoubleOpsWrapper<'_, T>
+where
+    T: DoubleOps,
+{
+    #[inline]
+    fn eq(&self, other: &Self) -> bool {
+        self.0.eq(other.0)
+    }
+}
+
+impl<T> Eq for DoubleOpsWrapper<'_, T> where T: DoubleOps {}
+
+impl<T> PartialOrd for DoubleOpsWrapper<'_, T>
+where
+    T: DoubleOps,
+{
+    #[inline]
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl<T> Ord for DoubleOpsWrapper<'_, T>
+where
+    T: DoubleOps,
+{
+    #[inline]
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.0.cmp(other.0)
+    }
+}
+
+impl<T> Hash for DoubleOpsWrapper<'_, T>
+where
+    T: DoubleOps,
+{
+    #[inline]
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.0.hash(state);
     }
 }
 

--- a/conjure-test/src/test/objects.rs
+++ b/conjure-test/src/test/objects.rs
@@ -483,3 +483,30 @@ fn set_of_objects_with_doubles() {
         .build();
     test_serde(&value, json);
 }
+
+#[test]
+fn map_double_values() {
+    let json = r#"
+    {
+        "raw": {
+            "foo": 1.0
+        },
+        "optional": {
+            "foo": 1.0,
+            "bar": null
+        },
+        "list": {
+            "foo": [
+                1.0
+            ]
+        }
+    }
+    "#;
+    let value = MapDoubleValues::builder()
+        .insert_raw("foo", 1.)
+        .insert_optional("foo", Some(1.))
+        .insert_optional("bar", None)
+        .insert_list("foo", vec![1.])
+        .build();
+    test_serde(&value, json);
+}

--- a/conjure-test/test-ir.json
+++ b/conjure-test/test-ir.json
@@ -24,8 +24,8 @@
       "type" : {
         "type" : "reference",
         "reference" : {
-          "name": "EmptyObject",
-          "package": "com.palantir.conjure"
+          "name" : "EmptyObject",
+          "package" : "com.palantir.conjure"
         }
       }
     } ],
@@ -474,6 +474,70 @@
           }
         }
       }
+    }
+  }, {
+    "type" : "object",
+    "object" : {
+      "typeName" : {
+        "name" : "MapDoubleValues",
+        "package" : "com.palantir.conjure"
+      },
+      "fields" : [ {
+        "fieldName" : "raw",
+        "type" : {
+          "type" : "map",
+          "map" : {
+            "keyType" : {
+              "type" : "primitive",
+              "primitive" : "STRING"
+            },
+            "valueType" : {
+              "type" : "primitive",
+              "primitive" : "DOUBLE"
+            }
+          }
+        }
+      }, {
+        "fieldName" : "optional",
+        "type" : {
+          "type" : "map",
+          "map" : {
+            "keyType" : {
+              "type" : "primitive",
+              "primitive" : "STRING"
+            },
+            "valueType" : {
+              "type" : "optional",
+              "optional" : {
+                "itemType" : {
+                  "type" : "primitive",
+                  "primitive" : "DOUBLE"
+                }
+              }
+            }
+          }
+        }
+      }, {
+        "fieldName" : "list",
+        "type" : {
+          "type" : "map",
+          "map" : {
+            "keyType" : {
+              "type" : "primitive",
+              "primitive" : "STRING"
+            },
+            "valueType" : {
+              "type" : "list",
+              "list" : {
+                "itemType" : {
+                  "type" : "primitive",
+                  "primitive" : "DOUBLE"
+                }
+              }
+            }
+          }
+        }
+      } ]
     }
   }, {
     "type" : "object",

--- a/conjure-test/test.yml
+++ b/conjure-test/test.yml
@@ -149,6 +149,11 @@ types:
       SetOfObjectsWithDoubles:
         fields:
           set: set<AllRequiredFields>
+      MapDoubleValues:
+        fields:
+          raw: map<string, double>
+          optional: map<string, optional<double>>
+          list: map<string, list<double>>
     errors:
       SimpleError:
         namespace: Test

